3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,271 @@
+// TODO(alanorozco): Refactor src/error.reportError so it does not contain big
+// transitive dependencies and can be included here.
+setReportError(() => {});
+
+/** @private @const {string} */
+const TAG_ = 'ampanalytics-lib';
+
+/** @private @const {number} */
+const MAX_QUEUE_SIZE_ = 100;
+
+/**
+ * Receives messages bound for this cross-domain iframe, from all creatives
+ */
+class AmpAnalytics3pMessageRouter {
+
+  /** @param {!Window} win */
+  constructor(win) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    let frameNameData = null;
+    try {
+      frameNameData = JSON.parse(this.win_.name);
+      if (!frameNameData.sentinel) {
+        throw new Error('JSON parsed, but did not include sentinel.');
+      }
+      // The sentinel is required by iframe-messaging-client, and is used to
+      // uniquely identify the frame as part of message routing
+      /** @const {string} */
+    } catch (e) {
+      dev().warn(TAG_, 'Unable to set sentinel from name attr: ' +
+        this.win_.name + ': ' + e.message);
+      return;
+    }
+    this.sentinel_ = frameNameData && frameNameData.sentinel;
+
+    /**
+     * Multiple creatives on a page may wish to use the same type of
+     * amp-analytics tag. This object provides a mapping between the
+     * IDs which identify which amp-analytics tag a message is to/from,
+     * with each ID's corresponding AmpAnalytics3pCreativeMessageRouter,
+     * which is an object that handles messages to/from a particular creative.
+     * @private {!Object<string, !AmpAnalytics3pCreativeMessageRouter>}
+     */
+    this.creativeMessageRouters_ = {};
+
+    /**
+     * Handles communication between frames
+     * @private {!IframeMessagingClient}
+     */
+    this.iframeMessagingClient_ = new IframeMessagingClient(win);
+    this.iframeMessagingClient_.setSentinel(this.sentinel_);
+    this.iframeMessagingClient_.registerCallback(
+      AMP_ANALYTICS_3P_MESSAGE_TYPE.CREATIVE,
+      messageContainer => {
+        dev().assert(messageContainer.data,
+          'Received empty new creative message');
+        Object.entries(messageContainer.data).forEach(entry => {
+          const creativeId = entry[0];
+          const extraData = entry[1];
+          if (this.creativeMessageRouters_.hasOwnProperty(creativeId)) {
+            dev().warn(TAG_, 'Duplicate new creative message for ' + creativeId);
+          }
+          this.creativeMessageRouters_[creativeId] =
+            this.creativeMessageRouters_[creativeId] ||
+            new AmpAnalytics3pCreativeMessageRouter(this, creativeId,
+              extraData);
+        });
+      });
+    this.iframeMessagingClient_.registerCallback(
+      AMP_ANALYTICS_3P_MESSAGE_TYPE.EVENT,
+      messageContainer => {
+        Object.entries(messageContainer.data).forEach(entry => {
+          const creativeId = entry[0];
+          const messages = entry[1];
+          dev().assert(messages.length > 0, 'Received empty events list');
```

Comment:
Done.
I agree/knew about the equivalence (given that we know it can't be negative)...I just think it looks more readable with the > 0. But will try to break myself of the habit.